python/lsst/daf/butler/_limited_butler.py

```diff
@@ -91,6 +95,135 @@ def putDirect(self, obj: Any, ref: DatasetRef) -> DatasetRef:
 
+    def _findDatasetRef(
+        self,
+        datasetRefOrType: DatasetRef | DatasetType | str,
+        dataId: DataId | None = None,
+        *,
+        collections: Any = None,
+        allowUnresolved: bool = False,
+        **kwargs: Any,
+    ) -> DatasetRef:
+        """Convert the standard get parameters to a `DatasetRef`
+
+        datasetRefOrType : `DatasetRef`, `DatasetType`, or `str`
+            Either a dataset reference, a dataset type, or the name of
+            a dataset type.
+            For the default implementation of a `LimitedButler`, the only
+            acceptable parameter is a resolved `DatasetRef`.
+        dataId : `dict` or `DataCoordinate`
+            A `dict` of `Dimension` link name, value pairs that label the
+            `DatasetRef` within a Collection. When `None`, a `DatasetRef`
+            should be provided as the first argument.
+        collections : Any, optional
+            Collections to be searched, overriding ``self.collections``.
+            Can be any of the types supported by the ``collections`` argument
+            to butler construction. Not used in a `LimitedButler`.
+        allowUnresolved : `bool`, optional
+            If `True`, return an unresolved `DatasetRef` if finding a resolved
+            one in the `Registry` fails.  Defaults to `False`.
+            For `LimitedButler` must always be `False`.
+        **kwargs
+            Additional keyword arguments used to augment or construct a
+            `DataCoordinate`.  See `DataCoordinate.standardize`
+            parameters. Not used in a `LimitedButler`.
+
+        Returns
+        -------
+        ref : `DatasetRef`
+            A resolved `DatasetRef`.
+
+        Raises
+        ------
+        AmbiguousDatasetError
+            Raised if an unresolved `DatasetRef` is passed as an input.
+        ValueError
+            Raised if an unresolved `DatasetRef` is passed as an input.
+
+        Notes
+        -----
+        Should be subclassed if a Butler needs to do a query using the
+        dataId and collections parameters.
+        """
+        if allowUnresolved:
+            raise ValueError("For LimitedButler allowUnresolved must always be False.")
+        if not isinstance(datasetRefOrType, DatasetRef):
+            raise ValueError("A LimitedButler can only retrieve datasets associated with a DatasetRef.")
+        if datasetRefOrType.id is None:
+            raise AmbiguousDatasetError(
+                f"Dataset of type {datasetRefOrType.datasetType.name} with "
+                f"data ID {datasetRefOrType.dataId} is not resolved."
+            )
+        return datasetRefOrType
+
+    def get(
+        self,
+        datasetRefOrType: DatasetRef | DatasetType | str,
+        dataId: DataId | None = None,
+        *,
+        parameters: dict[str, Any] | None = None,
+        collections: Any = None,
+        storageClass: StorageClass | str | None = None,
+        **kwargs: Any,
```

Comment:
I was thinking we'd just make the `LimitedButler` signature
```
def get(self, ref: DatasetRef, /, *, parameters: dict[str, Any] | None = None) -> Any:
```
Maybe it should have `storageClass`, too, but since `ref` has one in it, and QG execution can easily make sure `ref` has the right one in it, it doesn't seem necessary (and as we've seen, we can't assume the `ref` storage class is the Registry one).

That would also let us shrink the `_findDatasetRef` implementation (or just merge it into other methods), and it'd keep the `LimitedButler` interface from being misleading.

To make that work with MyPy we might have to use `typing.overload` in full `Butler` and make `datasetRefOrType` a positional-only parameter, to show that it does satisfy the original signature as well as adding a new one.

Same comment for `getDeferred`.

Reply:
If `LimitedButler.get()` has a different signature to `Butler.get()` that seems a bit odd to me since it implies that the user needs to understand that `Butler` is special and is the only subclass that is allowed to pass in something other than a `DatasetRef`. Don't we generally frown upon subclasses having completely different parameters to other subclasses?

Reply:
The user needs to understand that `Butler` is special regardless; I think the only thing that's actually up for debate here is whether _MyPy_ should understand that that `LimitedButler` requires a `DatasetRef` and hence `Butler` is special.

> Don't we generally frown upon subclasses having completely different parameters to other subclasses?

Yes, in the absence of overloading.  And overloading is maybe sort of frowned upon, but it's what users said they wanted, and once you accept overloading, we're not changing the signature; we're allowing a new one, and that is more analogous to adding a new method.

Reply:
@TallJimbo as you surmised, mypy has no problem with the subclass having additional parameters so long as the signature includes a `/` to force it not to support a keyword value usage.